### PR TITLE
SOLAR support - module EVM-IMP-SPECS

### PR DESCRIPTION
**Pull request by denis-bogdanas**
Module proofs does not depend on driver or web3, and is a new top-level module. Hence changes to Makefile.

Question: better name options?

Fixes: runtimeverification/firefly#283

### CODE REVIEW

**Comment by denis-bogdanas** [comment]
All passes now.

File: specs.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
+Statements for imperative proofs (SOLAR-style)
+====================
```
Comment (ehildenb) [inline specs.md]: Formatting. Also, call this evm-imp-specs.md.

File: specs.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+```k
+requires "evm.k"
+
+module EVM-SPECS
```
Comment (ehildenb) [inline specs.md]: `EVM-IMP-SPECS`

File: specs.md (hunk 1 above)
Comment (ehildenb) [reply specs.md]: The formatting should have the `===` extend the full length of the previous line.

File: specs.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    configuration
+      <kevm-specs>
+        <kevm/>
+        <commandVars> .Map </commandVars>
+      </kevm-specs>
+
+// Core specification scripting language
+// ======================================================================
+
```
Comment (ehildenb) [inline specs.md]: End the code block, have this be a proper level two header:

```
Core Specification Scripting Language
-------------------------------------
```

then start the next code block.

File: specs.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    syntax EthereumCommand ::= "#assert" Exp [strict]
+                             | "#assertFailure" Exp
+    rule #assert R:Bool => .
+      requires R
+
+    rule #assert R:Bool => #assertFailure R
+      requires notBool R
+
+syntax EthereumCommand ::= "#assume" Exp [strict]
```
Comment (ehildenb) [inline specs.md]: indent

File: specs.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+// Core specification scripting language
+// ======================================================================
+
+    syntax EthereumCommand ::= "#assert" Exp [strict]
+                             | "#assertFailure" Exp
+    rule #assert R:Bool => .
+      requires R
+
+    rule #assert R:Bool => #assertFailure R
+      requires notBool R
```
Comment (ehildenb) [inline specs.md]: These rules only apply in the `<k>` cell, so wrap them in `<k>` tags.

File: specs.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+          ~> #mkCall CALLER_ID ACCT_ID ACCT_ID PARSEDCODE 0 ARGS false
+          ~> #takeHalt "failure"
+          ~> #saveOutput V_SAVEOUT_ID
+        ...</k>
+        <acctID> ACCT_ID </acctID>
+        <code> PARSEDCODE </code>
+
+    //Dummy command at the beginning of <k> to ensure execution doesn't start with a spec rule 
+    syntax EthereumCommand ::= "#dummy"
+    rule #dummy => .
```
Comment (ehildenb) [inline specs.md]: `<k> ... </k>` tags.

File: specs.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+
+// Configuration access commands
+// ======================================================================
+
```
Comment (ehildenb) [inline specs.md]: end code block, level 2 header, start codeblock

File: specs.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    syntax EthereumCommand ::= "#saveRefund" Id
+    rule <k> #saveRefund X => . ...</k>
+         <refund> R </refund>
+         <commandVars> VARS => VARS[X <- R] </commandVars>
+
+
+// Specifications expression language
+// ======================================================================
+
```
Comment (ehildenb) [inline specs.md]: End code block, level 2 header, start code block

File: specs.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    syntax KResult ::= ByteArray | StatusCode | List | Map
+
+    syntax Exp ::= Exp "==S"  Exp  [seqstrict]
+                 | Exp "=/=S" Exp  [seqstrict]
+                 | Exp "+List" Exp [seqstrict]
+                 | KResult
+
+    rule R1:KResult ==S  R2:KResult => R1 ==K  R2
+    rule R1:KResult =/=S R2:KResult => R1 =/=K R2
+    rule L1:List +List L2:List => L1 L2
```
Comment (ehildenb) [inline specs.md]: `<k> ... </k>` tags.

File: specs.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    // Boolean expressions
+    syntax Exp ::= Exp "&&S" Exp [seqstrict, left]
+                 > Exp "||S" Exp [seqstrict, left]
+
+    rule R1:Bool &&S R2:Bool => R1 andBool R2
+    rule R1:Bool ||S R2:Bool => R1 orBool R2
```
Comment (ehildenb) [inline specs.md]: `<k> ... </k>` tags

File: specs.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+         <refund> R </refund>
+
+    rule <k> #getStorage(ACCT_ID) => S ...</k>
+         <acctID> ACCT_ID </acctID>
+         <storage> S </storage>
+
+    rule <k> #var(X) => VARS[X] ...</k>
+         <commandVars> VARS </commandVars>
+
+    rule #sizeWordStackExp(WS:WordStack) => #sizeWordStack(WS)
```
Comment (ehildenb) [inline specs.md]: `<k> ... </k>` tags.

**Comment by ehildenb** [comment]
Is there any reason to have this in evm-semantics repo and not in VSC repo?

File: Makefile [hunk 12]
```diff
@@ -144,26 +144,30 @@ $(PLUGIN_SUBMODULE)/make.timestamp:
...
+export SPECS_DEFN_FILE := evm-imp-specs
 
-k_files       := driver.k data.k network.k evm.k evm-types.k json.k krypto.k edsl.k evm-node.k web3.k asm.k state-loader.k serialization.k
+k_files       := driver.k data.k network.k evm.k evm-types.k json.k krypto.k edsl.k evm-node.k web3.k asm.k state-loader.k serialization.k evm-imp-specs.k
 EXTRA_K_FILES += $(MAIN_DEFN_FILE).k
 ALL_K_FILES   := $(k_files) $(EXTRA_K_FILES)
 
 llvm_dir    := $(DEFN_DIR)/llvm
 java_dir    := $(DEFN_DIR)/java
+specs_dir    := $(DEFN_DIR)/specs
```
Comment (ehildenb) [inline Makefile]:
```suggestion
specs_dir   := $(DEFN_DIR)/specs
```

File: evm-imp-specs.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+      requires R
+
+    rule <k> #assert R:Bool => #assertFailure R ...</k>
+      requires notBool R
+
+    syntax EthereumCommand ::= "#assume" Exp [strict]
+
+    //Adapted from driver.md `failure`
+    syntax EthereumCommand ::= "#takeHalt" String
+ // -------------------------------------------------------
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
 // ---------------------------------------------
```

File: evm-imp-specs.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+      </kevm-specs>
+```
+
+Core specification language
+---------------------------
+
+```k
+    syntax EthereumCommand ::= "#assert" Exp [strict]
+                             | "#assertFailure" Exp
+    rule <k> #assert R:Bool => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #assert R:Bool => . ... </k>
```

File: evm-imp-specs.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+      requires R
+
+    rule <k> #assert R:Bool => #assertFailure R ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #assert R:Bool => #assertFailure R ... </k>
```

File: evm-imp-specs.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    rule <k> #halt ~> #takeHalt _ => . ... </k>
+
+    //Shortcut for #mkCall and required setup
+    syntax EthereumCommand ::= "#mkCallShortcut" Id Int Int ByteArray Id
+    rule <k> #mkCallShortcut V_CONF_BACKUP CALLER_ID ACCT_ID ARGS V_SAVEOUT_ID
+          => #restoreEthereum V_CONF_BACKUP
+          ~> #mkCall CALLER_ID ACCT_ID ACCT_ID PARSEDCODE 0 ARGS false
+          ~> #takeHalt "failure"
+          ~> #saveOutput V_SAVEOUT_ID
+        ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
         ...
         </k>
```

File: evm-imp-specs.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        <acctID> ACCT_ID </acctID>
+        <code> PARSEDCODE </code>
+
+    //Dummy command at the beginning of <k> to ensure execution doesn't start with a spec rule 
+    syntax EthereumCommand ::= "#dummy"
+    rule <k> #dummy => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #dummy => . ... </k>
```

File: evm-imp-specs.md [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+Configuration access commands
+-----------------------------
+
+```k
+    syntax EthereumCommand ::= "#saveEthereum" Id
+    rule <k> #saveEthereum X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #saveEthereum X => . ... </k>
```

File: evm-imp-specs.md [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+         <ethereum> ETH </ethereum>
+         <commandVars> VARS => VARS[X <- ETH] </commandVars>
+
+    syntax EthereumCommand ::= "#restoreEthereum" Id
+    rule <k> #restoreEthereum X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #restoreEthereum X => . ... </k>
```

File: evm-imp-specs.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+         (<ethereum> _ </ethereum> => <ethereum> ETH </ethereum>)
+         <commandVars>... X |-> ETH ...</commandVars>
+
+    syntax EthereumCommand ::= "#saveOutput" Id
+    rule <k> #saveOutput X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #saveOutput X => . ... </k>
```

File: evm-imp-specs.md [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+         <output> OUT </output>
+         <commandVars> VARS => VARS[X <- OUT] </commandVars>
+
+    syntax EthereumCommand ::= "#saveStorage" Int Id
+    rule <k> #saveStorage ACCT_ID X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #saveStorage ACCT_ID X => . ... </k>
```

File: evm-imp-specs.md [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+         <acctID> ACCT_ID </acctID>
+         <storage> S </storage>
+         <commandVars> VARS => VARS[X <- S] </commandVars>
+
+    syntax EthereumCommand ::= "#saveLog" Id
+    rule <k> #saveLog X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #saveLog X => . ... </k>
```

File: evm-imp-specs.md [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+         <log> L </log>
+         <commandVars> VARS => VARS[X <- L] </commandVars>
+
+    syntax EthereumCommand ::= "#saveRefund" Id
+    rule <k> #saveRefund X => . ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #saveRefund X => . ... </k>
```

File: evm-imp-specs.md [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+    rule <k> R1:KResult ==S  R2:KResult => R1 ==K  R2 ...</k>
+    rule <k> R1:KResult =/=S R2:KResult => R1 =/=K R2 ...</k>
+    rule <k> L1:List +List L2:List      => L1 L2      ...</k>
+
+    // Boolean expressions
+    syntax Exp ::= Exp "&&S" Exp [seqstrict, left]
+                 > Exp "||S" Exp [seqstrict, left]
+
+    rule <k> R1:Bool &&S R2:Bool => R1 andBool R2 ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> R1:Bool &&S R2:Bool => R1 andBool R2 ... </k>
```

File: evm-imp-specs.md [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+    rule <k> R1:Bool ||S R2:Bool => R1 orBool R2  ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> R1:Bool ||S R2:Bool => R1 orBool R2  ... </k>
```

File: evm-imp-specs.md [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    // Configuration access and other helpers
+    syntax Exp ::= "#getStatusCode"
+                 | "#getOutput"
+                 | "#getLog"
+                 | "#getRefund"
+                 | #getStorage ( Int )
+                 | #var( Id )
+                 | #sizeWordStackExp ( Exp ) [strict]
+
+    rule <k> #getStatusCode => SC ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #getStatusCode => SC ... </k>
```

File: evm-imp-specs.md [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+         <statusCode> SC </statusCode>
+
+    rule <k> #getOutput => OUT ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #getOutput => OUT ... </k>
```

File: evm-imp-specs.md [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+         <output> OUT </output>
+
+    rule <k> #getLog => L ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #getLog => L ... </k>
```

File: evm-imp-specs.md [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+         <log> L </log>
+
+    rule <k> #getRefund => R ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #getRefund => R ... </k>
```

File: evm-imp-specs.md [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+         <refund> R </refund>
+
+    rule <k> #getStorage(ACCT_ID) => S ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #getStorage(ACCT_ID) => S ... </k>
```

File: evm-imp-specs.md [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+         <acctID> ACCT_ID </acctID>
+         <storage> S </storage>
+
+    rule <k> #var(X) => VARS[X] ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #var(X) => VARS[X] ... </k>
```

File: evm-imp-specs.md [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+         <commandVars> VARS </commandVars>
+
+    rule <k> #sizeWordStackExp(WS:WordStack) => #sizeWordStack(WS) ...</k>
```
Comment (ehildenb) [inline evm-imp-specs.md]:
```suggestion
    rule <k> #sizeWordStackExp(WS:WordStack) => #sizeWordStack(WS) ... </k>
```

**Comment by ehildenb** [comment]
I added some formatting changes. We use the same styling throughout the repo, I don't see any reason to have this file be any different.

Need some simple tests of this functionality before we can merge it, both as an example of how to run/use it and to make sure we don't break it too horribly on updates.

**Comment by denis-bogdanas** [comment]
@ehildenb ready for review

File: tests/specs/imp-specs/imp-specs-abstract-semantics.k [hunk 33]
```diff
@@ -0,0 +1,10 @@
+requires "evm-imp-specs.k"
+
+module IMP-SPECS-ABSTRACT-SEMANTICS
+  imports EVM-IMP-SPECS
+
+  rule #assume R:Bool => .
```
Comment (ehildenb) [inline tests/specs/imp-specs/imp-specs-abstract-semantics.k]:
```suggestion
  rule <k> #assume R:Bool => . ... </k>
```

**Pull request merged**